### PR TITLE
notebooks: Migrate query inputs from Monaco to CodeMirror

### PR DESCRIPTION
**Pull request by fkling**
This commit migrates the Monaco instances used by Notebooks as query inputs to use CodeMirror instead. It also refactors parts of the existing CodeMirror query input code to make it more reusable.

I'm sorry for the big diff, I'll point out the most important changes:

All instances use `<CodeMirrorQueryInput />` which provides the basic logic for syntax highlighting and token information (but not autocompletion or any specific keybindings). Example:

```
                <CodeMirrorQueryInput
                    value={queryInput}
                    isLightTheme={isLightTheme}
                    onEditorCreated={onEditorCreatedLocal}
                    patternType={SearchPatternType.literal}
                    interpretComments={true}
                    extensions={useMemo(
                        () => [blockKeymap({ runBlock }), changeListener(setQueryInput), singleLine, extension ?? []],
                        [setQueryInput, runBlock, extension]
                    )}
                />
```

A couple of extensions have been created/exported to make creating query inputs with the desired behavior easier:
- `blockKeymap`: Notebooks specific extension for defining `Mod+Enter` and `Escape` keybindings.
- `singleLine` exported from the core CodeMirror query input code to easily create single line query inputs.
- `editorHeight` was created to allow specifying a fixed and/or max editor height
- `createDefaultSuggestions` was updated to allow specifying which completions should be supported.

Since autocompletion/intelligence logic is shared between CodeMirror instances, the autocompletion logic is initialized in the respective files (`NotebookQueryBlock`, `NotebookFileBlockInputs`, `NotebookSymbolInputs`)

A reference to the editor instance is now directly maintained within the `SearchTypeSuggestionsInput` (vs it being passed down from the parent).

The example filter buttons used to update to notify the parent component about the value change and then update the text selection after the component re-rendered. Now the order is reversed: The buttons directly update the editor input (including text selection), which automatically triggers notifying the parent component.

I consolidated `setQueryInput` and `debouncedSetQueryInput`. It looks like it was only needed to make the example filter buttons apply immediately but they do that now due to inverting the flow.

Preventing CodeMirror from treating the focus `Enter` keypress as input can be more easily done by prevent the default event behavior.

## Test plan

Integration tests and manual inspection.

## App preview:

- [Web](https://sg-web-fkling-notebook-codemirror.onrender.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by fkling** [comment]
@novoselrok Since you know much more about Notebook behavior, I'd appreciate it if could test this locally to see if I have missed anything.

File: client/search-ui/src/input/extensions/completion.ts [hunk 1]
```diff
@@ -227,136 +227,151 @@ export function createDefaultSuggestionSources(options: {
...
-            // Default to the current cursor position (e.g. if the token is a
-            // whitespace, we want the suggestion to be inserted after it)
-            let from = context.position
-
-            if (token?.type === 'pattern') {
-                // If the token is a pattern (e.g. the start of a filter name),
-                // we want the suggestion to complete that name.
-                from = token.range.start
-            }
+    const sources: SuggestionSource<CompletionResult | null, SuggestionContext>[] = []
```
Comment (fkling) [inline client/search-ui/src/input/extensions/completion.ts:233]: The only difference in this file is rearranging the source creation logic to make it conditional. The actual source creation logic did not change.

File: client/web/src/notebooks/blocks/NotebookBlock.module.scss [hunk 2]
```diff
@@ -21,20 +21,3 @@
...
-    }
-
-    :global(.monaco-mouse-cursor-text) {
-        /* Hack: remove focus ring around Monaco editor cursor */
-        &:focus-within,
-        &:focus-visible {
-            box-shadow: none;
-        }
-    }
-}
```
Comment (fkling) [inline client/web/src/notebooks/blocks/NotebookBlock.module.scss:40]: This didn't seems needed. I couldn't see any box-shadow effect with CodeMirror.

File: client/web/src/notebooks/notebook/NotebookComponent.tsx [hunk 3]
```diff
@@ -331,25 +328,6 @@ export const NotebookComponent: React.FunctionComponent<React.PropsWithChildren<
         )
         useNotebookEventHandlers(notebookEventHandlersProps)
 
-        const sourcegraphSearchLanguageId = useQueryIntelligence(fetchStreamSuggestions, {
```
Comment (novoselrok) [inline client/web/src/notebooks/notebook/NotebookComponent.tsx:334]: So happy to see this go away 😄

File: client/web/src/notebooks/blocks/file/NotebookFileBlockInputs.tsx [hunk 4]
```diff
@@ -95,14 +112,13 @@ export const NotebookFileBlockInputs: React.FunctionComponent<
...
                 id={id}
-                editor={editor}
-                setEditor={setEditor}
                 label="Find a file using a Sourcegraph search query"
                 queryPrefix="type:path"
                 fetchSuggestions={fetchFileSuggestions}
                 countSuggestions={countSuggestions}
                 renderSuggestions={renderSuggestions}
-                {...props}
+                extension={useMemo(() => [queryCompletion, editorAttributes], [queryCompletion])}
```
Comment (novoselrok) [inline client/web/src/notebooks/blocks/file/NotebookFileBlockInputs.tsx:120]: I haven't tried this yet, so I'm not sure about using `useMemo` directly as a prop. I've only seen it assigned to a variable and then passed as a prop, but I guess this should work as well?

**Review by novoselrok — APPROVED** [review]
LGTM, I've checked it out locally and I'm trusting the integration tests to do most of the heavy lifting 😄

File: client/web/src/notebooks/blocks/file/NotebookFileBlockInputs.tsx (hunk 4 above)
Comment (fkling) [reply client/web/src/notebooks/blocks/file/NotebookFileBlockInputs.tsx:120]: Yeah, the only important thing is that hooks are called directly in the component (or in other hooks) and that they are always called in the same order.

File: client/web/src/notebooks/notebook/NotebookComponent.tsx (hunk 3 above)
Comment (fkling) [reply client/web/src/notebooks/notebook/NotebookComponent.tsx:334]: I can somewhat understand that it makes sense to share these kinds of configurations globally but in our case it's more of a hindrance.

**Pull request merged**